data_structures/bloom_filter.cpp

```diff
@@ -0,0 +1,282 @@
+    std::size_t size();
+    void add(std::size_t x);
+    bool contains(std::size_t x);
+};
+
+/**
+ * @brief Returns size of inner array
+ *
+ * @return size of inner array
+ */
+std::size_t Bitset::size() { return data.size(); }
+
+/**
+ * @brief Constructor for Bitset
+ *
+ * @param initSize amount of blocks, each contain sizeof(std::size_t) bits
+ */
+Bitset::Bitset(std::size_t initSize) : data(initSize) {}
+
+/**
+ * @brief Turn bit on position x to 1s
+ *
+ * @param x position to turn bit on
+ * @return void
+ */
+void Bitset::add(std::size_t x) {
+    std::size_t blockIndex = x / blockSize;
+    if (blockIndex >= data.size()) {
+        data.resize(blockIndex + 1);
+    }
+    data[blockIndex] |= 1 << (x % blockSize);
+}
+
+/**
+ * @brief Doest bitset contains element x
+ *
+ * @param x position in bitset to check
+ * @return true if: bit on position x is 1
+ * @return false if: bit on position x is 0
+ */
+bool Bitset::contains(std::size_t x) {
+    std::size_t blockIndex = x / blockSize;
+    if (blockIndex >= data.size()) {
+        return false;
+    }
+    return data[blockIndex] & (1 << (x % blockSize));
+}
+
+/**
+ * @brief Bloom filter template class
+ *
+ * @tparam T type, that we need to hash
+ */
+template <typename T>
+class BloomFilter {
+ private:
+    Bitset set;  ///< inner bitset for elements
+    std::vector<std::function<std::size_t(T)>>
+        hashFunks;  ///< hash functions for T type
+
+ public:
+    BloomFilter(std::size_t size,
+                std::initializer_list<std::function<std::size_t(T)>> funks);
+    void add(T x);
+    bool contains(T x);
+};
+
+/**
+ * @brief Constructor for Bloom filter
+ *
+ * @tparam T type, that we need to hash
+ * @param size initial size of Bloom filter
+ * @param funks hash functions for T type
+ */
+template <typename T>
+BloomFilter<T>::BloomFilter(
+    std::size_t size,
+    std::initializer_list<std::function<std::size_t(T)>> funks)
+    : set(size), hashFunks(funks) {}
+
+/**
+ * @brief Add function for Bloom filter
+ *
+ * @tparam T type, that we need to hash
+ * @param x element to add to filter
+ * @return void
+ */
+template <typename T>
+void BloomFilter<T>::add(T x) {
+    for (std::size_t i = 0; i < hashFunks.size(); i++) {
+        set.add(hashFunks[i](x) % (sizeof(std::size_t) * set.size()));
+    }
+}
+
+/**
+ * @brief Check element function for Bloom filter
+ *
+ * @tparam T type, that we need to hash
+ * @param x element to check in filter
+ * @return true if: element probably appears in filter
+ * @return false if: element certainly does not appear in filter
+ */
+template <typename T>
+bool BloomFilter<T>::contains(T x) {
+    for (std::size_t i = 0; i < hashFunks.size(); i++) {
+        if (set.contains(hashFunks[i](x) %
+                         (sizeof(std::size_t) * set.size())) == false) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * @brief Test for bitset
+ *
+ * @return void
+ */
+static void test_bitset() {
```

Comment:
Please move this function before the `main` function.


__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix

Reply:
Done